File: tui/tui.cpp
#include <algorithm>
#include <chrono>
#include <ctime>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <thread>
#include <utility>
#include <vector>

#define WIN32_LEAN_AND_MEAN
#include <windows.h>

namespace {

constexpr const char* kCursorHome = "\x1b[H";
constexpr const char* kClearScreen = "\x1b[2J";
constexpr const char* kHideCursor = "\x1b[?25l";
constexpr const char* kShowCursor = "\x1b[?25h";
constexpr const char* kResetStyle = "\x1b[0m";
constexpr const char* kBlueBackgroundCyanText = "\x1b[44m\x1b[96m";

bool EnableAnsiProcessing() {
    HANDLE handle = GetStdHandle(STD_OUTPUT_HANDLE);
    if (handle == INVALID_HANDLE_VALUE || handle == nullptr) {
        return false;
    }

    DWORD mode = 0;
    if (GetConsoleMode(handle, &mode) == 0) {
        return false;
    }

    return SetConsoleMode(handle, mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING) != 0;
}

std::pair<int, int> GetTerminalSize() {
    CONSOLE_SCREEN_BUFFER_INFO info{};
    HANDLE handle = GetStdHandle(STD_OUTPUT_HANDLE);
    if (handle != INVALID_HANDLE_VALUE && handle != nullptr && GetConsoleScreenBufferInfo(handle, &info) != 0) {
        const int width = info.srWindow.Right - info.srWindow.Left + 1;
        const int height = info.srWindow.Bottom - info.srWindow.Top + 1;
        return {std::max(width, 4), std::max(height, 4)};
    }

    return {80, 24};
}

void PlaceCenteredText(std::vector<std::string>& lines, int row, const std::string& text) {
    if (row <= 0 || row >= static_cast<int>(lines.size()) - 1) {
        return;
    }

    std::string& line = lines[row];
    const int innerWidth = static_cast<int>(line.size()) - 2;
    const std::string trimmed = text.substr(0, static_cast<std::size_t>(innerWidth));
    const int start = 1 + std::max(0, (innerWidth - static_cast<int>(trimmed.size())) / 2);

    for (std::size_t index = 0; index < trimmed.size(); ++index) {
        line[static_cast<std::size_t>(start) + index] = trimmed[index];
    }
}

std::string BuildScreen(int width, int height, const std::string& caption, const std::string& body) {
    width = std::max(width, 4);
    height = std::max(height, 4);
    const int innerWidth = width - 2;

    std::vector<std::string> lines;
    lines.reserve(static_cast<std::size_t>(height));
    lines.emplace_back("+" + std::string(static_cast<std::size_t>(innerWidth), '-') + "+");

    for (int row = 0; row < height - 2; ++row) {
        lines.emplace_back("|" + std::string(static_cast<std::size_t>(innerWidth), ' ') + "|");
    }

    lines.emplace_back("+" + std::string(static_cast<std::size_t>(innerWidth), '-') + "+");

    PlaceCenteredText(lines, 1, caption);
    PlaceCenteredText(lines, height / 2, body);

    std::ostringstream screen;
    screen << kBlueBackgroundCyanText;
    for (std::size_t index = 0; index < lines.size(); ++index) {
        screen << lines[index];
        if (index + 1 < lines.size()) {
            screen << '\n';
        }
    }
    screen << kResetStyle;
    return screen.str();
}

std::string CurrentTimeString() {
    const auto now = std::chrono::system_clock::now();
    const std::time_t currentTime = std::chrono::system_clock::to_time_t(now);

    std::tm localTime{};
    localtime_s(&localTime, &currentTime);

    std::ostringstream output;
    output << std::put_time(&localTime, "%H:%M:%S");
    return output.str();
}

void RestoreConsole() {
    std::cout << kResetStyle << kShowCursor << kCursorHome;
    std::cout.flush();
}

BOOL WINAPI ConsoleHandler(DWORD signal) {
    switch (signal) {
        case CTRL_C_EVENT:
        case CTRL_BREAK_EVENT:
        case CTRL_CLOSE_EVENT:
        case CTRL_LOGOFF_EVENT:
        case CTRL_SHUTDOWN_EVENT:
            RestoreConsole();
            return FALSE;
        default:
            return FALSE;
    }
}

class ConsoleSession {
public:
    ConsoleSession() {
        EnableAnsiProcessing();
        SetConsoleCtrlHandler(ConsoleHandler, TRUE);
        std::cout << kClearScreen << kCursorHome << kHideCursor;
        std::cout.flush();
    }

    ~ConsoleSession() {
        RestoreConsole();
    }

    ConsoleSession(const ConsoleSession&) = delete;
    ConsoleSession& operator=(const ConsoleSession&) = delete;
};

}  // namespace

int main() {
    ConsoleSession consoleSession;
    std::pair<int, int> lastSize{-1, -1};

    while (true) {
        const auto [width, height] = GetTerminalSize();
        const std::string screen = BuildScreen(width, height, "TUI Demo", CurrentTimeString());

        const bool resized = lastSize.first != width || lastSize.second != height;
        std::cout << (resized ? std::string(kClearScreen) + kCursorHome : std::string(kCursorHome));
        std::cout << screen;
        std::cout.flush();

        lastSize = {width, height};

        const auto now = std::chrono::system_clock::now();
        const auto nextTick = std::chrono::time_point_cast<std::chrono::seconds>(now) + std::chrono::seconds(1);
        std::this_thread::sleep_until(nextTick);
    }
}

File: tui/tui.py
import shutil
import sys
import time
from datetime import datetime


CURSOR_HOME = "\x1b[H"
CLEAR_SCREEN = "\x1b[2J"
HIDE_CURSOR = "\x1b[?25l"
SHOW_CURSOR = "\x1b[?25h"
RESET_STYLE = "\x1b[0m"
BG_BLUE = "\x1b[44m"
FG_CYAN = "\x1b[96m"

def place_centered_text(lines: list[list[str]], row: int, text: str) -> None:
	if row <= 0 or row >= len(lines) - 1:
		return

	inner_width = len(lines[row]) - 2
	trimmed_text = text[:inner_width]
	start = 1 + max(0, (inner_width - len(trimmed_text)) // 2)
	for index, char in enumerate(trimmed_text):
		lines[row][start + index] = char


def colorize_line(line: list[str], row: int, height: int) -> str:
	colored_parts: list[str] = []
	for index, char in enumerate(line):
		style = BG_BLUE + FG_CYAN
		colored_parts.append(f"{style}{char}")
	colored_parts.append(RESET_STYLE)
	return "".join(colored_parts)


def build_screen(width: int, height: int, caption: str, body: str) -> str:
	width = max(width, 4)
	height = max(height, 4)
	inner_width = width - 2

	lines = [list("+" + "-" * inner_width + "+")]
	for _ in range(height - 2):
		lines.append(list("|" + " " * inner_width + "|"))
	lines.append(list("+" + "-" * inner_width + "+"))

	place_centered_text(lines, 1, caption)
	place_centered_text(lines, height // 2, body)

	return "\n".join(colorize_line(line, row, height) for row, line in enumerate(lines))


def render_loop() -> None:

	last_size: tuple[int, int] | None = None

	sys.stdout.write(CLEAR_SCREEN + CURSOR_HOME + HIDE_CURSOR)
	sys.stdout.flush()

	try:
		while True:
			width, height = shutil.get_terminal_size(fallback=(80, 24))
			clock_text = datetime.now().strftime("%H:%M:%S")
			screen = build_screen(width, height, "TUI Demo", clock_text)

			prefix = CURSOR_HOME
			if last_size != (width, height):
				prefix = CLEAR_SCREEN + CURSOR_HOME
			sys.stdout.write(prefix + screen)
			sys.stdout.flush()


			last_size = (width, height)
			time.sleep(max(0.05, 1 - (time.time() % 1)))
	except KeyboardInterrupt:
		pass
	finally:
		sys.stdout.write(RESET_STYLE + SHOW_CURSOR + CURSOR_HOME)
		sys.stdout.flush()


if __name__ == "__main__":
	render_loop()
